webgpu.c

```diff
@@ -758,9 +758,50 @@ WGPUCommandEncoder wgpuDeviceCreateCommandEncoder(WGPUDevice device, WGPUCommand
 // {
 // }
 
-// WGPUShaderModule wgpuDeviceCreateShaderModule(WGPUDevice device, WGPUShaderModuleDescriptor const* descriptor)
-// {
-// }
+WGPUShaderModule wgpuDeviceCreateShaderModule(WGPUDevice device, WGPUShaderModuleDescriptor const* descriptor)
+{
+    if (!device || !descriptor) unreachable();
+
+    WGPUShaderSourceWGSL* wgsl_source = NULL;
+    switch (descriptor->nextInChain->sType) {
+        case WGPUSType_ShaderSourceWGSL:
+            wgsl_source = (WGPUShaderSourceWGSL*)descriptor->nextInChain;
+            break;
+        case WGPUSType_ShaderSourceSPIRV:
+            // consider adding SPIRV support
+            todo();
+            break;
+        default:
+            unreachable();
+    }
+
+    imports_string_t code_wasi = {
```

Comment:
I still think we should drop the +1's 

  .ptr = malloc(wgsl_source->code.length),
  .len = wgsl_source->code.length,
  memcpy(code_wasi.ptr, wgsl_source->code.data, wgsl_source->code.length);